### PR TITLE
Adding Pauli benchamrks

### PR DESCRIPTION
**Pull request by eliarbel**
### Summary
Benchmarking various Pauli & PauliList related methods

### CODE REVIEW

**Comment by eliarbel** [comment]
@chriseclectic can you please review the Pauli & PauliList benchmarks I've added ? Let me know if you think more methods should be included. 
Also, should we add SparsePauliOp benchmarks at this point or should we wait for its conversion to be PauliList-based first ?

File: test/benchmarks/quantum_info.py [hunk 1]
```diff
@@ -85,3 +87,66 @@ def time_compose(self, nqubits_length):
...
+            c1 = random_clifford(nqubits)
+            p1.evolve(c1)
+
+
+class PauliListBench:
+    def time_basic_ops(self):
+        nqubits = 500
+        length = 500
+
+        pl1 = PauliList([random_pauli(nqubits, True)
```
Comment (ikkoham) [inline test/benchmarks/quantum_info.py]: How about using `random_pauli_list`?

Comment (eliarbel) [reply test/benchmarks/quantum_info.py]: Yes, good idea, will do

Comment (eliarbel) [reply test/benchmarks/quantum_info.py]: Ikko, BTW why `random_pauli_list` is not in the import list of `quantum_info` ?

Comment (ikkoham) [reply test/benchmarks/quantum_info.py]: My bad. I'll add it soon.

File: test/benchmarks/quantum_info.py [hunk 2]
```diff
@@ -85,3 +88,95 @@ def time_compose(self, nqubits_length):
...
+    def time_basic_ops(self, nqubits):
+        iterations = 200
+        for _ in range(0, iterations):
+            p1 = random_pauli(nqubits, True)
+            p2 = random_pauli(nqubits, True)
+
+            p1.compose(p2)
+            p1.evolve(p2)  # by another Pauli, so by composition
+            p1.commutes(p2)
+            p1.to_instruction()
```
Comment (mtreinish) [inline test/benchmarks/quantum_info.py]: I know we discussed this before, but I'm still a bit concerned by the bundling of operations here. It makes it more difficult to isolate regressions. Like if this benchmark flags a regression on PR X how do we know if it's the conversion to an instruction, or composition that had the issue?

File: test/benchmarks/quantum_info.py [hunk 3]
```diff
@@ -85,3 +88,95 @@ def time_compose(self, nqubits_length):
...
         for i in range(length):
             cxdihedral.compose(self.random_cnotdihedral[i])
+
+
+class PauliBench:
+    params = [100, 200, 300, 400, 500]
+
+    def time_basic_ops(self, nqubits):
+        iterations = 200
+        for _ in range(0, iterations):
```
Comment (mtreinish) [inline test/benchmarks/quantum_info.py]: Normally we don't do the iteration like this in the benchmark and rely on asv to do multiple samples for us. You can tweak how it does this with attributes: https://asv.readthedocs.io/en/stable/benchmarks.html#timing-benchmarks

File: test/benchmarks/quantum_info.py [hunk 4]
```diff
@@ -85,3 +88,95 @@ def time_compose(self, nqubits_length):
...
+            p1 = random_pauli(nqubits, True)
+            p2 = random_pauli(nqubits, True)
```
Comment (mtreinish) [inline test/benchmarks/quantum_info.py]: This probably needs a seed for consistency between runs.

File: test/benchmarks/quantum_info.py [hunk 5]
```diff
@@ -85,3 +88,95 @@ def time_compose(self, nqubits_length):
         cxdihedral = CNOTDihedral(num_qubits=nqubits)
         for i in range(length):
             cxdihedral.compose(self.random_cnotdihedral[i])
+
+
+class PauliBench:
+    params = [100, 200, 300, 400, 500]
```
Comment (mtreinish) [inline test/benchmarks/quantum_info.py:94]: Can you also set a `param_names` attribute here so the graphs render this as num_qubits:

```suggestion
    params = [100, 200, 300, 400, 500]
    param_names = ["num_qubits"]
```

File: test/benchmarks/quantum_info.py [hunk 6]
```diff
@@ -85,3 +88,95 @@ def time_compose(self, nqubits_length):
...
+
+    def time_basic_ops(self, nqubits):
+        iterations = 200
+        for _ in range(0, iterations):
+            p1 = random_pauli(nqubits, True)
```
Comment (mtreinish) [inline test/benchmarks/quantum_info.py]: Does it make sense to split the random pauli generation out into a `setup()` method so we leave the timed portion to just the ops and not the random generation (which honestly will probably be mostly constrained on numpy's rng and system entropy).

**Review by mtreinish — COMMENTED** [review]
I left some comments on the first benchmark class but I think they apply to all the classes in the file

**Comment by eliarbel** [comment]
Thanks @mtreinish  for the review and comments. I've addressed almost all them. The only thing I've left as before is the use of unfixed random seed. I think it's better this way as otherwise we are left with benchmarking the same samples and potentially miss statistical chance of hitting a corner case. Given a known seed for each run, is there a way to log or publish it in ASV somehow in case we need to use it for debug ?

File: test/benchmarks/quantum_info.py (hunk 3 above)
Comment (eliarbel) [reply test/benchmarks/quantum_info.py]: Done, removed

File: test/benchmarks/quantum_info.py (hunk 4 above)
Comment (eliarbel) [reply test/benchmarks/quantum_info.py]: That one I would leave as is, to have some statistical coverage of the possible Pauli's. Theoretically we may find some corner case

File: test/benchmarks/quantum_info.py (hunk 6 above)
Comment (eliarbel) [reply test/benchmarks/quantum_info.py]: Those are very quick, but split I put it in `setup()` anyway.

File: test/benchmarks/quantum_info.py (hunk 2 above)
Comment (eliarbel) [reply test/benchmarks/quantum_info.py]: no problem, I split them into distinct methods. This will entail many graphs in the dashboard. If it will bother us we can revert.

**Review by mtreinish — APPROVED** [review]
LGTM, this looks good to me. I'm still a bit concerned about the seed but it hopefully shouldn't cause too much noise in the data. Tracing it back to a seed value though will be tricky. The only output we have is to stdout or stderr which will only get printed during the run and not get associated with the generated graphs. So tracing an outlier back will require manually looking at the logs and pulling the seed out, so it's probably not worth the effort. If the benchmarks are too noisy after a few data points we can always go in and change things.

I also tested these all locally and they're quite quick so it won't add any noticeable delay to our per commit benchmarking time.

**Comment by eliarbel** [comment]
Thanks @mtreinish !
